File: DLL/DLL.c
#include <stdio.h>
#include <stdlib.h>

struct DLL 
{
    int data;
    struct DLL *prev;
    struct DLL *next;
};

struct DLL *first = NULL;
struct DLL *last = NULL;

void insertAtBeg(int element) 
{
    struct DLL *NewNode = (struct DLL*)malloc(sizeof(struct DLL));
    NewNode->data = element;
    NewNode->prev = NULL;
    NewNode->next = first;

    if (first == NULL) 
    {
        first = last = NewNode;
    } 
    else 
    {
        first->prev = NewNode;
        first = NewNode;
    }
    printf("%d WAS INSERTED AT THE BEGINNING\n", element);
}

void insertAtEnd(int element) 
{
    struct DLL *NewNode = (struct DLL*)malloc(sizeof(struct DLL));
    NewNode->data = element;
    NewNode->next = NULL;
    NewNode->prev = last;

    if (first == NULL) 
    {
        first = last = NewNode;
    } 
    else 
    {
        last->next = NewNode;
        last = NewNode;
    }
    printf("%d WAS INSERTED AT THE END\n", element);
}

void insertAtPosition(int element, int pos) 
{
    struct DLL *temp = first;
    int i;
    if (pos <= 0 || first == NULL) 
    {
        printf("INVALID INSERTION!\n");
        return;
    }
    if (pos == 1) 
    {
        insertAtBeg(element);
        return;
    }

    for (i = 1; i < pos - 1 && temp != NULL; i++) 
        temp = temp->next;

    if (temp == NULL || temp == last) 
    {
        insertAtEnd(element);
    } 
    else 
    {
        struct DLL *NewNode = (struct DLL*)malloc(sizeof(struct DLL));
        NewNode->data = element;

        NewNode->next = temp->next;
        NewNode->prev = temp;
        if (temp->next != NULL) 
        {
            temp->next->prev = NewNode;
        }
        temp->next = NewNode;

        printf("%d WAS INSERTED AT POSITION %d\n", element, pos);
    }
}

void deletionFromBeg() 
{
    struct DLL *temp = first;
    if (first == NULL) 
    {
        printf("EMPTY LIST!\n");
        return;
    }
    printf("%d IS GOING TO BE DELETED\n", temp->data);

    if (first == last) 
    { 
        first = last = NULL;
    } else 
    {
        first = first->next;
        first->prev = NULL;
    }
    free(temp);
}

void deletionFromEnd() 
{
    struct DLL *temp = last;
    if (last == NULL) 
    {
        printf("EMPTY LIST!\n");
        return;
    }
    printf("%d IS GOING TO BE DELETED\n", temp->data);

    if (first == last) 
    { 
        first = last = NULL;
    } 
    else 
    {
        last = last->prev;
        last->next = NULL;
    }
    free(temp);
}

void deletionFromPosition(int pos) 
{
    struct DLL *temp = first;
    if (first == NULL || pos <= 0) 
    {
        printf("INVALID DELETION!\n");
        return;
    }

    if (pos == 1) 
    {
        deletionFromBeg();
        return;
    }

    int i;
    for (i = 1; i < pos && temp != NULL; i++) 
    {
        temp = temp->next;
    }

    if (temp == NULL) 
    {
        printf("INVALID POSITION!\n");
        return;
    }

    printf("%d IS GOING TO BE DELETED\n", temp->data);

    if (temp == last) 
    {
        deletionFromEnd();
        return;
    }

    if (temp->prev != NULL)
        temp->prev->next = temp->next;

    if (temp->next != NULL)
        temp->next->prev = temp->prev;

    free(temp);
}

void display() 
{
    struct DLL *temp = first;

    if (first == NULL) 
    {
        printf("EMPTY LIST!\n");
        return;
    }

    while (temp != NULL) 
    {
        printf("%d <-> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

void search(int key) 
{
    if (first == NULL) 
    {
        printf("EMPTY LIST!\n");
        return;
    }
    struct DLL *temp = first;
    while (temp != NULL) 
    {
        if (temp->data == key) 
        {
            printf("Element Found!\n");
            return;
        }
        temp = temp->next;
    }
    printf("SEARCH UNSUCCESSFUL!\n");
}

int main() 
{
    int choice, subchoice, element, pos;

    while (1) 
    {
        printf("\n--- MENU ---\n");
        printf("1. Insert\n2. Delete\n3. Search\n4. Display\n5. Exit\n");

        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) 
        {
        case 1:
            printf("Insert:\n1. At Beginning\n2. At End\n3. At Specific Position\n");
            printf("Enter your choice: ");
            scanf("%d", &subchoice);

            printf("Enter element to insert: ");
            scanf("%d", &element);

            if (subchoice == 1)
                insertAtBeg(element);

            else if (subchoice == 2)
                insertAtEnd(element);

            else if (subchoice == 3) 
            {
                printf("Enter position: ");
                scanf("%d", &pos);
                insertAtPosition(element, pos);
            } 
            else
                printf("Invalid insert choice!\n");
            break;

        case 2:
            printf("Delete:\n1. From Beginning\n2. From End\n3. From Specific Position\n");
            printf("Enter your choice: ");
            scanf("%d", &subchoice);

            if (subchoice == 1)
                deletionFromBeg();

            else if (subchoice == 2)
                deletionFromEnd();

            else if (subchoice == 3) 
            {
                printf("Enter position: ");
                scanf("%d", &pos);
                deletionFromPosition(pos);
            } 
            else
                printf("Invalid delete choice!\n");
            break;

        case 3:
            printf("Enter element to search: ");
            scanf("%d", &element);
            search(element);
            break;

        case 4:
            display();
            break;

        case 5:
            exit(0);

        default:
            printf("Invalid choice. Try again.\n");
        }
    }

    return 0;
}

File: LinkedLists/DLL/DLL.py
def create_node(data):
    return {'data': data, 'next': None, 'prev': None}

first = None
last = None

def insertAtBeg(element):
    global first, last
    new_node = create_node(element)

    if first is None:
        first = last = new_node

    else:
        new_node['next'] = first
        first['prev'] = new_node
        first = new_node

    print(f"{element} WAS INSERTED AT THE BEGINNING")

def insertAtEnd(element):
    global first, last
    new_node = create_node(element)

    if first is None:
        first = last = new_node

    else:
        last['next'] = new_node
        new_node['prev'] = last
        last = new_node

    print(f"{element} WAS INSERTED AT THE END")

def insertAtPosition(element, pos):
    global first, last

    if first is None or pos <= 0:
        print("INVALID INSERTION!")
        return
    
    if pos == 1:
        insertAtBeg(element)
        return

    new_node = create_node(element)
    temp = first
    i = 1

    while i < pos - 1 and temp is not None:
        temp = temp['next']
        i += 1

    if temp is None or temp == last:
        insertAtEnd(element)
        
    else:
        new_node['next'] = temp['next']
        new_node['prev'] = temp

        if temp['next'] is not None:
            temp['next']['prev'] = new_node

        temp['next'] = new_node
        print(f"{element} WAS INSERTED AT POSITION {pos}")

def deletionFromBeg():
    global first, last

    if first is None:
        print("EMPTY LIST!")

    elif first['next'] is None:
        print(f"{first['data']} IS GOING TO BE DELETED")
        first = last = None

    else:
        print(f"{first['data']} IS GOING TO BE DELETED")
        first = first['next']
        first['prev'] = None

def deletionFromEnd():
    global first, last

    if first is None:
        print("EMPTY LIST!")

    elif first['next'] is None:
        print(f"{first['data']} IS GOING TO BE DELETED")
        first = last = None

    else:
        print(f"{last['data']} IS GOING TO BE DELETED")
        last = last['prev']
        last['next'] = None

def deletionFromPosition(pos):

    global first, last

    if first is None or pos <= 0:
        print("INVALID DELETION!")
        return
    
    if pos == 1:
        deletionFromBeg()
        return
    
    temp = first
    i = 1
    
    while i < pos and temp is not None:
        temp = temp['next']
        i += 1

    if temp is None:
        print("INVALID POSITION!")
        return
    
    if temp == last:
        deletionFromEnd()

    else:
        print(f"{temp['data']} IS GOING TO BE DELETED")

        if temp['prev'] is not None:
            temp['prev']['next'] = temp['next']

        if temp['next'] is not None:
            temp['next']['prev'] = temp['prev']

def display():
    global first

    if first is None:
        print("EMPTY LIST!")
        return
    
    temp = first

    while temp is not None:
        print(f"{temp['data']} <-> ", end="")
        temp = temp['next']

    print("NULL")

def search(key):
    global first

    if first is None:
        print("EMPTY LIST!")
        return
    
    temp = first
    found = False

    while temp is not None:
        if temp['data'] == key:
            found = True
            break

        temp = temp['next']

    print("Element Found!" if found else "SEARCH UNSUCCESSFUL!")


def main():
    while True:
        print("\n--- MENU ---")
        print("1. Insert\n2. Delete\n3. Search\n4. Display\n5. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            print("Insert:\n1. At Beginning\n2. At End\n3. At Specific Position")
            subchoice = int(input("Enter your choice: "))
            element = int(input("Enter element to insert: "))

            if subchoice == 1:
                insertAtBeg(element)

            elif subchoice == 2:
                insertAtEnd(element)

            elif subchoice == 3:
                pos = int(input("Enter position: "))
                insertAtPosition(element, pos)

            else:
                print("Invalid insert choice!")

        elif choice == 2:
            print("Delete:\n1. From Beginning\n2. From End\n3. From Specific Position")
            subchoice = int(input("Enter your choice: "))

            if subchoice == 1:
                deletionFromBeg()

            elif subchoice == 2:
                deletionFromEnd()
                
            elif subchoice == 3:
                pos = int(input("Enter position: "))
                deletionFromPosition(pos)

            else:
                print("Invalid delete choice!")

        elif choice == 3:
            key = int(input("Enter element to search: "))
            search(key)

        elif choice == 4:
            display()

        elif choice == 5:
            print("Exiting program.")
            break

        else:
            print("Invalid choice. Try again.")


if __name__ == "__main__":
    main()
